device.go

```diff
@@ -22,34 +24,67 @@ import (
 )
 
 const (
-	driver9pType  = "9p"
-	driverBlkType = "blk"
+	driver9pType   = "9p"
+	driverBlkType  = "blk"
+	driverSCSIType = "scsi"
+)
+
+// SCSI variables
+var (
+	// Here in "0:0", the first number is the SCSI host number because
+	// only one SCSI controller has been plugged, while the second number
+	// is always 0.
+	scsiHostChannel = "0:0:"
+	sysClassPrefix  = "/sys/class"
+	scsiDiskPrefix  = filepath.Join(sysClassPrefix, "scsi_disk", scsiHostChannel)
+	scsiBlockSuffix = "block"
+	scsiDiskSuffix  = filepath.Join("/device", scsiBlockSuffix)
+	scsiHostPath    = filepath.Join(sysClassPrefix, "scsi_host")
 )
 
 type deviceDriversHandler func(device pb.Device, spec *pb.Spec) error
 
 var deviceDriversHandlerList = map[string]deviceDriversHandler{
-	driverBlkType: deviceBlockDeviceDriverHandler,
+	driverBlkType:  deviceBlkBlockDeviceDriverHandler,
+	driverSCSIType: deviceSCSIBlockDeviceDriverHandler,
 }
 
-func deviceBlockDeviceDriverHandler(device pb.Device, spec *pb.Spec) error {
+func deviceBlkBlockDeviceDriverHandler(device pb.Device, spec *pb.Spec) error {
 	// First need to make sure the expected device shows up properly,
 	// and then we need to retrieve its device info (such as major and
 	// minor numbers), useful to update the device provided
 	// through the OCI specification.
-	if err := waitForDevice(device.VmPath); err != nil {
+	devName := strings.TrimPrefix(device.VmPath, devPrefix)
+	checkUevent := func(uEv *uevent.Uevent) bool {
+		return (uEv.Action == "add" &&
+			filepath.Base(uEv.DevPath) == devName)
+	}
+	if err := waitForDevice(device.VmPath, devName, checkUevent); err != nil {
+		return err
+	}
+
+	return updateSpecDeviceList(device, spec)
+}
+
+func deviceSCSIBlockDeviceDriverHandler(device pb.Device, spec *pb.Spec) error {
+	// Retrieve the device path from SCSI address.
+	devPath, err := getSCSIDevPath(device.Id)
+	if err != nil {
 		return err
 	}
+	device.VmPath = devPath
+
+	return updateSpecDeviceList(device, spec)
+}
 
+func updateSpecDeviceList(device pb.Device, spec *pb.Spec) error {
```

Comment:
Can you add documentation comments for this function as well.

Reply:
Yes sure.